### ISSUE TITLE
SNI based dynamic certificate loading support

### ISSUE THREAD

**Issue by p-id** [varnish/hitch#117]
Hitch at present loads all required certificates to serve different domains at the initialization time and would require configuration files to be reloaded even if a single certificate is to updated.

The proposal is to add support in Hitch to utilize an external service to fetch a valid certificate for specific (sub)domain on receiving a hostname via SNI hook.

Few additional points to be considered
1. The external service can fetch the certificate from entity like Let's Encrypt
2. Certificate once fetched is made available in cache (for all worker processes)
3. A service management hook to purge/clear cached certificate without reload

**Comment by lkarsten** [comment]
Hi. Sorry for not replying earlier.

I don't fully understand what this would accomplish. Can you explain in a bit more depth?

Currently we have the keys, certificates and any intermediate CAs on disk in a file per SNI name. They are referenced by a pem-file configuration entry (or startup argument), and reloaded on SIGHUP. Only the files on disk that are changed since last reload are loaded, to avoid the slow creation of openssl ssl contexts. New worker children are forked, and the old ones told to finish they connections and shut down.

How do you envision the interface to this external service?

Is it the external service that persists these files for the next time, or should Hitch do that part?

**Comment by p-id** [comment]
I agree that this is indeed a rather exotic feature request which may not be applicable in scenarios where we have a few (less than 10 domains or so) served via Hitch as an SSL off-loader 

However, there are couple of benefits with this approach about utilising an external Certificate delivery service (let's term it as Ex-CSD)

1. The actual domain to be serviced is kind of run-time configurable [no restart/reload] with Hitch relying on Ex-CSD to ensure if the domain is serviceable and furnish the associated front-end configuration certificates.
2. Once the certificate is loaded via Ex-CSD call, the certificates (SSL-Contexts) are cached and served till a PURGE request is send across to stop servicing that particular (sub) domain or a reload/restart to clear complete cache.
3. The interface in intended to have all parameters as in the Hitch front-end configuration, though the format could be either Json or a serialised 'C' FrontEnd struct binary blob. 
4. No certificates/files are stored on disk at-least from Hitch perspective, the Ex-CSD service can in theory enforce all desired kind of remote certificate fetch mechanisms.

If the above points do seem inline with an intended add-on feature to Hitch, I am all open to work on it and have a pull-request latest by end of year, if not before.

**Comment by p-id** [comment]
In the absence of any known standard interface for requesting/revoking existing certificates, the option most appropriate to Hitch would be the configuration file format for frontend definition used in Hitch.

The basic process flow would be quite similar to

1. Hitch starts up with spawning 'N' worker process and pre-loading a static set of SSL certificates

2.On receiving a hostname in SNI callback, an initial search is done in the pre-loaded certificate lookup (uthash based lookup)

- 2.a If the certificate(SSL* context) is found in internal cache, the specific context is used.

- 2.b If no matching entry is found against the hostname then an external Certificate delivery service (let's term it as Ex-CSD) is contacted via a local SEQ_PACKET socket containing a datagram (hostname value as received in SNI callback). Ex-CSD service would look for an appropriate certificate and return back the effective configuration for the specific fronted (empty configuration if no matching certificate found)

3. Once the certificate is loaded via Ex-CSD call, the certificates (SSL-Contexts) are cached and served till a PURGE request is send across to stop servicing that particular (sub) domain or a reload/restart to clear complete cache. The certificate (frontend) purge request is received on a separate management (SEQ_PACKET) socket. The Ex-CSD service is responsible for sending out the PURGE request (in Hitch frontend-config format)

**Comment by FGasper** [comment]
Pure-FTPd has something that might fit here for dynamic SNI.
